src/lazy/expanded/macro_evaluator.rs

```diff
@@ -374,7 +415,9 @@ impl<'top, D: Decoder> MacroExpansion<'top, D> {
         match self.next_step()? {
             // If the expansion produces anything other than a final value, there's a bug.
             MacroExpansionStep::FinalStep(Some(ValueExpr::ValueLiteral(value))) => Ok(value),
-            _ => unreachable!("e-expression-backed lazy values must yield a single value literal"),
+            _ => ice!(IonResult::decoding_error(format!(
+                "expansion of {self:?} was required to produce exactly one value",
+            ))),
```

Comment:
This path is an optimization available to macros that are statically determined to expand to a single non-system value at template-compile time. Consider this macro definition:
```
(macro Point2D ($x $y)
  {
    x: $x,
    y: $y,
  }
)
```
When the `TemplateCompiler` processes this, it notices that the body of the macro is an unannotated struct. This means that it will always evaluate to exactly one value and that value cannot be a symbol table or encoding directive (because those require specific annotations).

The codebase refers to this category of macros as "singleton macros." They offer a few benefits:
* Because it will _definitely_ be exactly one value, the reader can hand out a `LazyValue` holding the e-expression as its backing data. Other macros cannot do this because if you're holding a `LazyValue` and the macro evaluates to nothing or 100 things, there's not a way for the application to handle those outcomes.
* At the top level, singleton macros can be skipped over without evaluation because we know they won't produce a system value that could modify the encoding context.
* Expanding a singleton macro doesn't require an evaluator with a stack because as soon as you've gotten a value, you're done--no need to `pop()` and preserve state.

Now that we're looking at this more closely though, the decoding error should actually be a `panic`--if it ever happens, it's a bug in the library and not a problem with the data. While I'm in there I'll also add some of the description above to the `ExpansionSingleton` type.